File: Library/src/main/java/org/endeavourhealth/library/model/DataModelProperty.java
package org.endeavourhealth.library.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonSetter;
import java.io.Serializable;
import org.endeavourhealth.library.model.tripletree.TTIriRef;

public class DataModelProperty implements Serializable {

  private TTIriRef property;
  private TTIriRef type;
  private String minInclusive;
  private String minExclusive;
  private String maxInclusive;
  private String maxExclusive;
  private String pattern;
  private TTIriRef inheritedFrom;
  private int order;

  public int getOrder() {
    return order;
  }

  public DataModelProperty setOrder(int order) {
    this.order = order;
    return this;
  }

  public TTIriRef getProperty() {
    return property;
  }

  public DataModelProperty setProperty(TTIriRef property) {
    this.property = property;
    return this;
  }

  @JsonIgnore
  public boolean hasProperty() {
    return property != null;
  }

  public String getPattern() {
    return pattern;
  }

  public DataModelProperty setPattern(String pattern) {
    this.pattern = pattern;
    return this;
  }

  public TTIriRef getType() {
    return type;
  }

  @JsonSetter
  public DataModelProperty setType(TTIriRef objectType) {
    this.type = objectType;
    return this;
  }

  @JsonIgnore
  public boolean hasType() {
    return type != null;
  }

  public String getMinInclusive() {
    return minInclusive;
  }

  public DataModelProperty setMinInclusive(String minInclusive) {
    this.minInclusive = minInclusive;
    return this;
  }

  public String getMinExclusive() {
    return minExclusive;
  }

  public DataModelProperty setMinExclusive(String minExclusive) {
    this.minExclusive = minExclusive;
    return this;
  }

  public String getMaxInclusive() {
    return maxInclusive;
  }

  public DataModelProperty setMaxInclusive(String maxInclusive) {
    this.maxInclusive = maxInclusive;
    return this;
  }

  public String getMaxExclusive() {
    return maxExclusive;
  }

  public DataModelProperty setMaxExclusive(String maxExclusive) {
    this.maxExclusive = maxExclusive;
    return this;
  }

  public TTIriRef getInheritedFrom() {
    if (null == inheritedFrom) return new TTIriRef();
    return inheritedFrom;
  }

  @JsonSetter
  public DataModelProperty setInheritedFrom(TTIriRef inheritedFrom) {
    this.inheritedFrom = inheritedFrom;
    return this;
  }

  @JsonIgnore
  public boolean isArray() {
    return maxExclusive == null || maxExclusive.isEmpty() || maxExclusive.equals("0");
  }
}

File: src/interfaces/DataModelProperty.ts
import z from "zod";

import { TTIriRef, TTIriRefSchema } from "./TTIriRef";

// export interface DataModelProperty {
//   property?: TTIriRef;
//   type?: TTIriRef;
//   minInclusive?: string;
//   minExclusive?: string;
//   maxInclusive?: string;
//   maxExclusive?: string;
//   pattern?: string;
//   inheritedFrom?: TTIriRef;
//   order?: number;
// }

export const DataModelPropertySchema = z.strictObject({
  property: TTIriRefSchema.optional(),
  type: TTIriRefSchema.optional(),
  minInclusive: z.string().optional(),
  minExclusive: z.string().optional(),
  maxInclusive: z.string().optional(),
  maxExclusive: z.string().optional(),
  pattern: z.string().optional(),
  inheritedFrom: TTIriRefSchema.optional(),
  order: z.number().optional()
});

export type DataModelProperty = z.output<typeof DataModelPropertySchema>;
